alica_engine/src/engine/RuntimePlanFactory.cpp

```diff
@@ -22,11 +22,21 @@ RuntimePlanFactory::RuntimePlanFactory(
 
 void RuntimePlanFactory::init(std::unique_ptr<IPlanCreator>&& pc)
 {
-    _creator = std::move(pc);
+    {
+        std::lock_guard lk(_m);
+        _creator = std::move(pc);
+        _initialized = true;
+    }
+    _cv.notify_all();
 }
 
 std::unique_ptr<BasicPlan> RuntimePlanFactory::create(int64_t id, const Plan* planModel) const
 {
+    std::unique_lock lk(_m);
+    while (!_initialized) {
+        _cv.wait(lk);
+    }
+
     PlanContext ctx{_globalBlackboard, planModel->getName(), planModel, _traceFactory};
     std::unique_ptr<BasicPlan> basicPlan = _creator->createPlan(id, ctx);
     if (!basicPlan) {
```

Comment:
Can probably unlock at this point.

Reply:
For what effect we should unlock at this point?
We need this sync because we starting communication before we complete whole initialization and we have cases when we have calls here before we receive _creator. If we somehow can get a recursive call here, I really want to know where

Reply:
Ah, now I see. Yep, I guess I could put all synchronization stuff in a nested scope to allow concurrent method invocation.